src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -338,7 +335,33 @@ private SendingReport sendEvents(EventsInfo info) throws IOException {
             search = events.size() == batchSize;
         }
 
-        return new SendingReport(CrawlerAction.NONE, interval, dataProcessorName, dataProcessorVersion, numberOfEvents, 0);
+        return new SendingReport(CrawlerAction.NONE, interval, dataProcessorName, dataProcessorVersion, numberOfEvents + diff, 0);
+    }
+
+    private EventResponse sendEventsToProcessor(DataProcessorService dataProcessor, EventDataRequest eventRequest) {
+        EventResponse response = dataProcessor.sendEvent(eventRequest);
+        metrics.processorMethodInvoked(ProcessorMethod.SEND_EVENT);
+        return response;
+    }
+
+    private Pair<@NotNull Integer, @Nullable EventData> processServiceResponse(@Nullable EventID checkpoint, List<EventData> events) {
+        if (checkpoint == null) {
+            return new Pair<>(events.size(), null);
+        }
+        int processed = 0;
+        EventData checkpointEvent = null;
+        for (EventData event : events) {
+            if (checkpointEvent == null) {
+                processed++;
+            }
+            if (event.getEventId().equals(checkpoint)) {
+                checkpointEvent = event;
+            }
+        }
+        if (!events.isEmpty()) {
+            metrics.lastEvent(events.get(events.size() - 1));
```

Comment:
Could we move this code close to the updateRecoveryState call to reflect the actual state of database

Reply:
This code does not suppose to reflect the state of the database. It should show the current crawler position. The RecoveryState might never be updated at all if the processor does not return any checkpoints